### PR TITLE
Improve Notebooks

### PR DESCRIPTION
**Pull request by MathieuBsqt**
Je suis repassé sur le notebook de découverte de YOLOv8. 
J'ai effectué plusieurs modifications. J'aurais peut-être du m'y prendre autrement pour les approuver ou non au cas par cas? Dans tous les cas, on pourra en discuter.

Voici les modifications effectuées et les raisons: 
- Modification des instructions de téléchargement du dataset (elles ont dû changer depuis la rédaction de ton tuto).
- Ajout d'un --quiet au unzip, sinon l'output est massive 
- Le lien du Swift Public traînait dans les 2 notebooks. Je l'ai supprimé du notebook de test, mais pas de la solution. Mais peut-être voulais-tu le conserver pour éviter aux gens d'aller sur Roboflow? J'ai un doute 
- Visualisation de la matrice de confusion que je trouve pas mal en + des graphs des métriques
- L'export des poids du modèle était fait un peu trop tôt (pas dans l'étape `Export trained weights for future inference`).
Le reste concerne l'ajout de descriptions / précisions

En plus de ces modfis, je souhaitais te faire quelques remarques:
- Pour le test du modèle sur des images, ça aurait été cool de visualiser directement les images résultantes depuis le notebook (plutôt que de devoir aller les chercher dans le dossier des résultats). Malheureusement, j'ai testé avec le paramètre show=True, mais ça n'a pas l'air de fonctionner :( 
- Peut-être dans la correction devrions-nous ajouter les liens des documentations de chaque méthode utilisée (afin que les utilisateurs voient comment on a trouvé la réponse, et quels paramètres ils peuvent utiliser. [Predict : https://docs.ultralytics.com/modes/predict/ ; Train: https://docs.ultralytics.com/modes/train/ ; Val: https://docs.ultralytics.com/modes/val/ ; Export: https://docs.ultralytics.com/modes/export/]
- Décrire davantage les paramètres utilisés (comme cela est fait pour la fonction train, avec le mention de device, ...) 
- Est-ce normal que les paramètres utilisés soient mentionnés dans la partie training du notebook (pour celui qui n'est pas le notebook solution) ?
- On pourrait laisser certaines images dans la correction (graphiques des métriques). 
- Il y a deux phrases qui m'ont l'air inexpliquées ("How to calculatge Precision and Recall" ; "How to calculate IoU ?"
- Dans les loss functions, je suis pas sûr de retrouver les mêmes que toi. Je dois creuser ce point.
- On pourrait ajouter une mention à l'oral / dans le notebook sur l'importance d'avoir un dataset correctement divisé (répartition train/test/val) et diversifié (couleurs de peau, tatouages, ...) et aussi de qualité (éclairage, qualité des images et leur grand nombre)
- Dans l'étape `Import dependencies and check GPU availability`, tu`import shutil, os`. Il me semble qu' `os` n'est pas utilisé, et `shutil` n'est pas obligatoire dans le notebook et n'intervient qu'à la fin. Je me demande s'il ne serait pas mieux d'importer shutil que lors de l'étape d'exportation, afin de ne pas faire croire à l'utilisateur qu'il n'a pas réussi la première étape du notebook, en oubliant d'importer cette librairie.

TODO: 
- Transfer learning / Fine-tuning
- Paramètres training (batch size, ...) & Dataset preprocessing ? 
- Loss functions 
- Détections en direct?

Suite aux réponses de Stéphane:
✅ Remettre le lien du dataset dans les 2 notebooks
Voir si je peux run sans import os
Import shutil à la fin 
Décrire davantages les steps / Inclure les liens des documentations (Essayer de faire en sorte qu'ils ne CTRL C V pas la solution mais cherchent pas eux-mêmes.
Voir ce qu'il entend par être sec et les paramètres
Enlever les paramètres qui sont expliqués dans le notebook classique (ctrl v venant de la solution)
Voir si on modifie les paramètres du training (batch size & Freeze) puis laisser les images dans la correction
"How to calculatge Precision and Recall" ; "How to calculate IoU ?"
Voir ce qu'il entend par "Yes, je le faisais en phase avec toi" & Préparer ce passage (voir si on fait une slide ou pas, car on les montre pas à ce moment là il me semble)

### CODE REVIEW

**Comment by MathieuBsqt** [comment]
**Concernant AI Training**:

✅ J'ai vérifié et le GPU est bien utilisé avec l'image de base `python:3.8` donc c'est nickel! 
❓Lors du passage de Notebooks à Training, les notebooks sont-ils stoppés? (afin de lancer la synchronisation du Swift qui est branché en volume et dans lequel on télécharge le dataset).
❓ Quel était le pull le plus rapide entre une image venant de Docker Hub vs Shared registry? 
😕 Lorsque je stream les logs de mon training avec `ovhai job logs -f`, je n'ai eu aucune information pendant les x epochs. J'ai l'impression que les logs ne sont affichés qu'une fois les x epochs terminés, et non epoch par epoch. C'est dommage, mais je ne pense pas qu'on y puisse grand chose. Pourtant dans le notebook, le training retournait bien des logs epochs par epochs. 

**Concernant AI Deploy**: 
💡 Peut-être devrions-nous préciser dans le commentaire du Dockerfile lié à l'instruction `CMD` que l'on souhaite run une application Streamlit? Même si l'on passe cette partie de construction d'images, j'ai peur que certaines personnes tentent de l'éxécuter directement avec Python.

**Comment by philippart-s** [comment]
> Je suis repassé sur le notebook de découverte de YOLOv8. J'ai effectué plusieurs modifications. J'aurais peut-être du m'y prendre autrement pour les approuver ou non au cas par cas? Dans tous les cas, on pourra en discuter.
> 
> Voici les modifications effectuées et les raisons:
> 
>     * Modification des instructions de téléchargement du dataset (elles ont dû changer depuis la rédaction de ton tuto).
👍 

>     * Ajout d'un --quiet au unzip, sinon l'output est massive
Merci ça me saoulais ! 

>     * Le lien du Swift Public traînait dans les 2 notebooks. Je l'ai supprimé du notebook de test, mais pas de la solution. Mais peut-être voulais-tu le conserver pour éviter aux gens d'aller sur Roboflow? J'ai un doute
Oui je voulais leur éviter de créer un compte, c'est pour cela que j'avais mis le lien dans le Notebook solution + de travail pour leur éviter cette étape qui apporte peu de valeur ajoutée je trouve 

>     * Visualisation de la matrice de confusion que je trouve pas mal en + des graphs des métriques
👍

>     * L'export des poids du modèle était fait un peu trop tôt (pas dans l'étape `Export trained weights for future inference`).
>       Le reste concerne l'ajout de descriptions / précisions
👍 
 
> En plus de ces modfis, je souhaitais te faire quelques remarques:
> 
>     * Pour le test du modèle sur des images, ça aurait été cool de visualiser directement les images résultantes depuis le notebook (plutôt que de devoir aller les chercher dans le dossier des résultats). Malheureusement, j'ai testé avec le paramètre show=True, mais ça n'a pas l'air de fonctionner :(
Ah oui c'est ballot c'est une bonne idée car ça évite de faire plein de cliques
 
>     * Peut-être dans la correction devrions-nous ajouter les liens des documentations de chaque méthode utilisée (afin que les utilisateurs voient comment on a trouvé la réponse, et quels paramètres ils peuvent utiliser. [Predict : https://docs.ultralytics.com/modes/predict/ ; Train: https://docs.ultralytics.com/modes/train/ ; Val: https://docs.ultralytics.com/modes/val/ ; Export: https://docs.ultralytics.com/modes/export/]
Carrément une bonne idée, je ne suis pas satisfais à l'heure actuelle de cette partie car ils ne sont pas assez autonomes pour des recherches et cela se traduit souvent par juste copier / coller la solution

 
>     * Décrire davantage les paramètres utilisés (comme cela est fait pour la fonction train, avec le mention de device, ...)
En phase, je suis un peu sec sur certains paramètres mais c'est une bonne idée
 
>     * Est-ce normal que les paramètres utilisés soient mentionnés dans la partie training du notebook (pour celui qui n'est pas le notebook solution) ?
Je pense qu'il y a du copié / merdé dans le descriptif 😅
 
>     * On pourrait laisser certaines images dans la correction (graphiques des métriques).
👍

>     * Il y a deux phrases qui m'ont l'air inexpliquées ("How to calculatge Precision and Recall" ; "How to calculate IoU ?"
> 
>     * Dans les loss functions, je suis pas sûr de retrouver les mêmes que toi. Je dois creuser ce point.
J'avoue que pour cette partie de l'ai copiée / collée depuis un notebook que l'on avait et je ne suis pas sûr qu'elle soit juste 

>     * On pourrait ajouter une mention à l'oral / dans le notebook sur l'importance d'avoir un dataset correctement divisé (répartition train/test/val) et diversifié (couleurs de peau, tatouages, ...) et aussi de qualité (éclairage, qualité des images et leur grand nombre)
Yes, je le faisais en phase avec toi
 
>     * Dans l'étape `Import dependencies and check GPU availability`, tu`import shutil, os`. Il me semble qu' `os` n'est pas utilisé, et `shutil` n'est pas obligatoire dans le notebook et n'intervient qu'à la fin. Je me demande s'il ne serait pas mieux d'importer shutil que lors de l'étape d'exportation, afin de ne pas faire croire à l'utilisateur qu'il n'a pas réussi la première étape du notebook, en oubliant d'importer cette librairie.  
oui c'est  un vieux reflex de dev, les imports tout en haut 😉 , mais en phase et pour le os à vérifier mais je pense que tu as raison

 
> 
> TODO:
> 
>     * Transfer learning / Fine-tuning
> 
>     * Paramètres training (batch size, ...) & Dataset preprocessing ?
> 
>     * Loss functions

**Comment by philippart-s** [comment]
> **Concernant AI Training**:
> 
> ✅ J'ai vérifié et le GPU est bien utilisé avec l'image de base `python:3.8` donc c'est nickel! 

👍

>❓Lors du passage de Notebooks à Training, les notebooks sont-ils stoppés? (afin de lancer la synchronisation du Swift qui est branché en volume et dans lequel on télécharge le dataset).
   
Le S3 est en cache donc je ne les stoppe pas, comme ça ils peuvent y revenir à loisir et le training/deplpy accèdent aux data tout de même

>❓ Quel était le pull le plus rapide entre une image venant de Docker Hub vs Shared registry?

De Docker Hub c'est constant entre 8 -10 mins, chez nous entre 12 - 25 mins !

> 😕 Lorsque je stream les logs de mon training avec `ovhai job logs -f`, je n'ai eu aucune information pendant les x epochs. J'ai l'impression que les logs ne sont affichés qu'une fois les x epochs terminés, et non epoch par epoch. C'est dommage, mais je ne pense pas qu'on y puisse grand chose. Pourtant dans le notebook, le training retournait bien des logs epochs par epochs.

J'ai constaté la même chose, je me le suis noté sur mon fichier de bugs / améliorations pour en parler à l'équipe qd on le parcourra.
En fait de temps en temps on voit le début mais le stream s'arrête

> 
> **Concernant AI Deploy**: 
✅ J'ai testé le modèle et il fonctionne super bien. Sans chercher à le piéger, je n'ai rencontré aucune erreur. 

👍

>💡 Peut-être devrions-nous préciser dans le commentaire du Dockerfile lié à l'instruction `CMD` que l'on souhaite run une application Streamlit? 

Même si l'on passe cette partie de construction d'images, j'ai peur que certaines personnes tentent de l'éxécuter directement avec Python. 
Oui tu as raison et on a jamais trop de doc ;)

>🤩 Pour pousser le truc encore plus loin, il nous faudrait (du temps déjà) pour créer un véritable jeu! Avec 2 inputs de camera par exemple, où tu te prends à tour de rôle en photo. Et le VS s'effectue entre les 2 inputs.

Tu ne crois pas si bien dire ;) : https://github.com/devrel-workshop/101-AI-and-py/issues/9
Je l'ai close car pas le temps / compétence pour le faire

**Comment by MathieuBsqt** [comment]
Les deux notebooks contiennent maintenant d'avantage d'explications, de docs. 
Il me reste à détailler l'analyse des graphiques et des fonctions de loss, vérifier que le modèle entraîné avec BatchSize=64 & Freeze 10 n'overfit pas et soit toujours utilisable.

J'ai également ajouté ces arguments (Batch & Freeze) dans le script solution train.py
